### ISSUE TITLE
Pluggable server-side storage engines for session data

### ISSUE THREAD

**Issue by zandbelt** [bungle/lua-resty-session#13]
I'd very much like to see server side support for session storage as listed in the roadmap. Any chance that this will move forward in the near future? FWIW: I'd be willing to contribute to the backends themselves (shm, files, memcached, redis...) but I don't feel comfortable to make the changes to the core infrastructure to start supporting those.

**Comment by bungle** [comment]
Okay, thanks for showing interest in this. I will start to look at it next week. I may need to open another branch for that where the work will be done. There is one pull-request that adds Redis support, but not in a pluggable fashion.

**Comment by bungle** [comment]
The storing is rather easy:

Shm would be something like this:

```lua
local shm = {}

function shm.read(id)
    return ngx.shared.sessions:get(id)
end

function shm.write(id, data, expiration)
    return ngx.shared.sessions:set(id, data, expiration)
end

return shm
```

The tricky part is to put locking as server side storages needs it, and possibly for read only and read write scenarios (e.g. read locks an read-write locks). The locking is not needed for cookie stored session data. Also server side storages need a lot more configuration (e.g. locking configs, shm dictionary name configs, redis host and port infos with unis socket support etc.).

We still need to send a cookie, and it is better to sign the cookie as well. We still need to encrypt data etc. So it seems this will introduce a lot of complexity that is not needed on cookie stored session data. Kinda hard to combine these aproaches. But I will try it. The performance of server side session storage will be probably less than the cookie stored session data as it needs locking.

**Comment by bungle** [comment]
Just a small update to this. I have now figured out somewhat working proposal to this. I have already made cookie and shm storage's and others should be relatively easy to add. There are a few things I need to figure out still until I put this in a public. I'm not sure if we can do this without breaking backward compability (it's rather close, though). So expect me to put something to the public in next few days. I would really appreciate if @zandbelt or someone else would like to create their own storage adapter to it then. So stay tuned.

**Comment by zandbelt** [comment]
great, I could add memcache support through https://github.com/openresty/lua-resty-memcached

**Comment by bungle** [comment]
Okay,

Here is my first take on pluggable storages and serializers (maybe later the encrypters as well) pushed on storages branch:
https://github.com/bungle/lua-resty-session/tree/storages

The code seems to be backward compatible. You can see the first two storage adapters here:

1. [Client Side Cookie](https://github.com/bungle/lua-resty-session/blob/storages/lib/resty/session/storage/cookie.lua)
2. [ngx.shared.DICT](https://github.com/bungle/lua-resty-session/blob/storages/lib/resty/session/storage/shm.lua)

Error handling and logging still needs some refirements, but overall it seems ok, and the API for pluggable adapters isn't too bad (somewhat strange in some points). What do you think?

**Comment by zandbelt** [comment]
looks great to me! will give it a try this week

**Comment by bungle** [comment]
Great,

Please don't hesitate to ask questions.

Quick and dirty guide:

Storage can be selected from nginx.conf:

```nginx
# cookie is the default storage
set $session_storage "shm";
```

Storage adapter needs to have at least these functions (and they need to reside on resty.session.storage.XXX):

```lua
-- creates a new adapter:
table XXX.new()
-- this opens session (only short lived locks here):
session_id, expires, enrypted_data, hmac XXX.open(cookie_string, session_lifetime)
-- this saves data and returns a new cookie_string (to be send to client),
-- if close argument is true you can release the lock
cookie_string XXX.save(session_id, expires, encrypted_data, hmac, close)
```

Optionally the adapter can implement:

```lua
-- called when read-write session is started (good place to acquire lock):
XXX.start(session_id)
-- called when session is destoyed (you should release lock here as well):
XXX.destroy(session_id)
```

Serialization and encryption / decryption and hmac checking and cookie setting is done the same way for all the adapters. The example shm adapter still stores expiration time and hmac inside the cookie (making it secure as the server side memory cannot be edited without changing client side hmac the same time and vise versa, and client side hmac doesn't have all the information as data resides on server). But this is not mandatory. I still recommend though to have some hmac mechanism to check that cookie was not tampered, and not trust the session id as the only thing.

About locking:

With shm adapter I used lua-resty-lock, and it works fine as long as we are using single nginx server or sticky sessions (shm doesn't work with multiple web frontends without sticky sessions). As both the locks and the data reside on the same address space. You cannot use lua-resty-lock if you want to implement memcached adapter and get rid of sticky sessions. You need to implement locking in memcached that all the servers can access. Locking needs to be implemented by session.id. We don't support more granular locks now (it is rather hard to lock on session.data.key level and more granular locking makes things rather complex and error prone). Please also note that all the locks you greate should timeout automatically if user forgets to call session:save() after he had called session:start() (shm adapter uses right now resty-lock default timeout / expiration settings: https://github.com/openresty/lua-resty-lock#new).

I think that the current API will work for memcached too.

**Comment by zandbelt** [comment]
haven't really looked in to it, but getting this error:

    2015/08/25 12:27:55 [error] 17276#0: *1 lua entry thread aborted: runtime error:   
    ...projects/nginx/install/lua/resty/session/storage/shm.lua:46: attempt to index field 'sessions' (a nil value)
    stack traceback:
    coroutine 0:
	    ...projects/nginx/install/lua/resty/session/storage/shm.lua: in function 'save'
	    ...s/hzandbelt/projects/nginx/install/lua/resty/session.lua:104: in function 'save'
	    ...s/hzandbelt/projects/nginx/install/lua/resty/session.lua:291: in function 'start'

any clue?

**Comment by bungle** [comment]
I think that you are missing shared dict.

Under http, please put these:

```nginx
lua_shared_dict sessions       1m;
lua_shared_dict sessions_locks 1m;
```

And under server you can select shm like this:

```nginx
set $session_storage "shm";
```

I think that this is what you are missing. There are also two other conf variables:

```nginx
set $session_shm_sessions "sessions";
set $session_shm_sessions_locks "sessions_locks";
```

That you can use to change the shared dict name that is used to store sessions. Those above are the defaults.

**Comment by zandbelt** [comment]
brilliant :-) all works for me now in a backwards compatible way; later this week I'll take a stab at the memcache plugin

I am wondering about requirement for locks though: since session identifiers differer between different browsers, locks are for the case where a user makes parallel updates to a session from a single browser, i.e. in different tabs? And then the global locking as implemented prevents tabs from being loaded in parallel but forces serial loading? Is that acceptable from a performance/behavioral point of view?

**Comment by bungle** [comment]
Yes, locking is for serializing the parallel access from the same session. It is mainly an issue with frames/iframes, and more so these days with Ajax. So it will not serialize different sessions but the requests with the same session (and when opening session as read-only by just using session.open(), the lock will be held only a small amount of time (but has to wait for previous read-write session to release the lock). The locking isn't needed on cookie based sessions. In certain cases you could probably run lock-less just fine if your code doesn't cause these kind of conflicts:

Without Locking:
![Without Locking](https://www.leaseweblabs.com/wp-content/uploads/2013/10/session-access-without-locking.png)

With Locking:
![With Locking](https://www.leaseweblabs.com/wp-content/uploads/2013/10/session-access-with-locking.png)

**Comment by bungle** [comment]
Okay, this is now in storages branch. Thank you very much. I will close this now. Next couple of weeks I will refine these a bit if I found something, and then this needs to be documented until we merge it in master branch and released as, maybe 1.8 or 2.0.
